src/dotnet/commands/dotnet-build/DotnetProjectBuilder.cs

```diff
@@ -0,0 +1,253 @@
+                {
+                    Directory.CreateDirectory(directoryName);
+                }
+                File.Copy(file, destFileName, true);
+            }
+        }
+
+        private void MakeRunnable(ProjectGraphNode graphNode)
+        {
+            var runtimeContext = graphNode.ProjectContext.ProjectFile.HasRuntimeOutput(_args.ConfigValue) ?
+                graphNode.ProjectContext.CreateRuntimeContext(_args.GetRuntimes()) :
+                graphNode.ProjectContext;
+
+            var outputPaths = runtimeContext.GetOutputPaths(_args.ConfigValue, _args.BuildBasePathValue, _args.OutputValue);
+            var libraryExporter = runtimeContext.CreateExporter(_args.ConfigValue, _args.BuildBasePathValue);
+
+            CopyCompilationOutput(outputPaths);
+
+            var executable = new Executable(runtimeContext, outputPaths, libraryExporter, _args.ConfigValue);
+            executable.MakeCompilationOutputRunnable();
+        }
+
+        protected override CompilationResult RunCompile(ProjectGraphNode projectNode)
+        {
+            try
+            {
+                var managedCompiler = new ManagedCompiler(_scriptRunner, _commandFactory);
+
+                var success = managedCompiler.Compile(projectNode.ProjectContext, _args);
+                if (projectNode.IsRoot)
+                {
+                    MakeRunnable(projectNode);
+                    PrintSummary(success);
+                }
+
+                return success ? CompilationResult.Success : CompilationResult.Failure;
+            }
+            finally
+            {
+                StampProjectWithSDKVersion(projectNode.ProjectContext);
+            }
+        }
+
+        private bool CLIChangedSinceLastCompilation(ProjectContext project)
+        {
+            var currentVersionFile = DotnetFiles.VersionFile;
+            var versionFileFromLastCompile = project.GetSDKVersionFile(_args.ConfigValue, _args.BuildBasePathValue, _args.OutputValue);
+
+            if (!File.Exists(currentVersionFile))
+            {
+                // this CLI does not have a version file; cannot tell if CLI changed
+                return false;
+            }
+
+            if (!File.Exists(versionFileFromLastCompile))
+            {
+                // this is the first compilation; cannot tell if CLI changed
+                return false;
+            }
+
+            var currentContent = ComputeCurrentVersionFileData();
+
+            var versionsAreEqual = string.Equals(currentContent, File.ReadAllText(versionFileFromLastCompile), StringComparison.OrdinalIgnoreCase);
+
+            return !versionsAreEqual;
+        }
+
+        protected override bool NeedsRebuilding(ProjectGraphNode  graphNode)
+        {
+            if (_args.ShouldNotUseIncrementality)
+            {
+                return true;
+            }
+            var preconditions = _preconditionManager.GatherIncrementalPreconditions(graphNode);
```

Comment:
Visitors aside, even though the args are the same, by taking the decision on each node rather than at the whole graph level, you can end up with a scenario where part of the graph passes the preconditions and part of the graph does not pass. This would be an incorrect state.

---
In reply to: [60173858](https://github.com/dotnet/cli/pull/2574#discussion_r60173858) [](ancestors = 60173858)